### PR TITLE
fix(ir): a column slice of a multi-row tile must not materialize over its own source

### LINKED ISSUE #2010: [Bug] t[:, a:b] on a multi-row tile silently returns wrong data — pl.tile.slice encodes a dense MemRef over its own live source
## Background

**`t[:, a:b]` — a column slice of a MULTI-ROW tile — silently returns wrong data.** It compiles
clean, raises nothing, produces no NaN. `ROWS = 1` is correct; every `ROWS >= 2` is corrupted.

This is about as basic as a DSL operation gets, and it is the fourth silent data-corruption defect
found in this same investigation (after #2003, #2007, #2006).

Found in qwen3-14b's fused decode (`models/qwen3/14b/decode_layer.py`), whose RoPE step column-slices
a `[16, 128]` tile into halves. The kernel returns wrong tokens; the in-repo golden cannot see it
because that golden is a second on-device dispatch of the same code path.

Reproduction environment:

| Component | Version |
|---|---|
| pypto | `8ebddcb8` (`main`) + the fixes for #2003 / #2007 / #2006 — reproduces with all three applied |
| simpler | `c94aa9f3` |
| ptoas | `0.48` |
| pto-isa | `83d01313` (the pin) |
| CANN | `9.0.0` |
| NPU | Ascend 910B (`a2a3`) |

### Component

IR / MemRef (`pl.tile.slice`)

## Root cause

pypto lowers `t[:, a:b]` to `pl.tile.slice` -> `MemRef(parent_buf, byte_offset = a * esize,
size = rows * cols_slice * esize)` — a **flat byte offset into the parent with a DENSE size**.
That encoding is only coherent when `rows == 1`.

Generated `.pto` for the qwen3 case (`q_heads` is `[16, 128]` f32, row pitch 512 B):

```mlir
%q_heads @76032  16x128 f32   (row pitch 512 B)
%q_lo    @76032  16x 64 f32   pad=0   (row pitch 256 B)   <-- base == q_heads
%q_hi    @76288  16x 64 f32   pad=0   (row pitch 256 B)   <-- base == q_heads + 256 B
pto.textract ins(%q_heads, 0,  0) outs(%q_lo)
pto.textract ins(%q_heads, 0, 64) outs(%q_hi)
```

`textract` must repack **strided (pitch 128) -> dense (pitch 64)**, and its destination lands
**on top of its own still-live source**:

- `q_lo`'s extract trashes `q_heads`;
- `q_hi`'s extract then reads the already-trashed `q_heads` **and** overwrites `q_lo` — which is read
  again afterwards.
- Only **row 0** survives, because for row 0 the dense destination address happens to equal its
  source address.

This was simulated byte-for-byte on the host from the kernel's own inputs: the simulation predicts
the device's `q_heads` **and** the downstream `q_rot` with **max|diff| = 0.000000 across all 2048
rows**.

**Why `ROWS = 1` is safe** (and why the qwen3 K path is unaffected while Q is destroyed): with one
row, source and destination pitches coincide, so the repack is an identity copy. qwen3's K path
row-slices to 1 row *first* (`k_normed[0:1, :]`) and only then column-slices; its Q path
column-slices the full 16-row tile. A Q/K bisect on the real kernel isolated the bug to Q exactly
along that line.

## Actual behaviour

```
   ROWS   max_abs_err     wrong elems    wrong rows  verdict
  --------------------------------------------------------------------
      1      0.000000       0/128          0/1       PASS
      2      3.602502      64/256          1/2       *** WRONG ***
      4      7.598506     128/512          2/4       *** WRONG ***
      8      7.048662     256/1024         4/8       *** WRONG ***
     16      9.420689     512/2048         8/16      *** WRONG ***
```

Signature at `ROWS = 16` — `out[0, 64:68]` comes back holding **row 1's lower half**:

```
device   out[0, 64:68]        = [-0.5881, -0.3706, -0.3759, -0.3649]
correct  x[0, 64:68] * w[:4]  = [-0.3789,  0.8139,  0.5189, -0.8323]
x[1, 0:4] * w[0:4]            = [-0.5881, -0.3706, -0.3759, -0.3649]   <-- what actually lands
```

Exactly the dense-repack-over-live-source signature.

## Steps to reproduce

Standalone; imports only `pypto.language` + `pypto.runtime`. No pypto-lib, no golden harness. ~100
lines. **No `pl.spmd`, no `pl.system.syncall`, no `pl.split_aiv`, no `pl.create_tensor`** — a column
slice of a multi-row tile is sufficient on its own.

```bash
python repro_colslice.py -d <device_id>     # exits 1 on the bug
```

<details>
<summary><b>repro_colslice.py</b></summary>

```python
"""STANDALONE repro — pypto only (pypto.language + pypto.runtime). No pypto-lib, no golden
harness.

A COLUMN slice of a MULTI-ROW tile is silently miscompiled.

    t   = x[0:ROWS, :]            # [ROWS, 128] f32 tile in UB, row pitch 128 floats
    lo  = t[:, 0:64]              # -> pto.textract into a DENSE [ROWS,64] tile (pitch 64)
    hi  = t[:, 64:128]            #    ALLOCATED AT t's OWN BASE ADDRESS (+0 / +64 floats)

pypto lowers each column slice to a `pto.textract` that repacks strided(pitch 128) ->
dense(pitch 64), and the memory allocator gives the destination the *subview's* address,
i.e. INSIDE the still-live source tile. The repack is therefore performed IN PLACE over its
own input:

  * lo's extract overwrites t (dest row r lands at t+64r, source row r sits at t+128r);
  * hi's extract then reads the ALREADY-TRASHED t, and its dest (t+64 .. t+64+64*ROWS)
    overwrites lo as well.

Only row 0 survives (its destination address equals its source address). For ROWS == 1 the
whole thing degenerates to an identity copy, so a single-row tile is unaffected — which is
why the same construct is harmless elsewhere.

Run:  python repro_colslice.py -d 3
"""
import argparse

import torch

import pypto.language as pl
from pypto.runtime import RunConfig

COLS = 128
HALF = 64

ap = argparse.ArgumentParser()
ap.add_argument("-d", "--device", type=int, default=3)
args = ap.parse_args()

import pypto.pypto_core as _c  # noqa: E402
print(f"[pypto] {_c.__file__}\n")

cfg = RunConfig(platform="a2a3", device_id=args.device, enable_dep_gen=False)


def build(rows):
    @pl.jit
    def k(
        x: pl.Tensor[[rows, COLS], pl.FP32],
        w: pl.Tensor[[1, HALF], pl.FP32],
        out: pl.Out[pl.Tensor[[rows, COLS], pl.FP32]],
    ):
        with pl.at(level=pl.Level.CORE_GROUP):
            t = x[0:rows, :]                    # [rows, 128] tile
            lo = t[:, 0:HALF]                   # column slice  <-- the defect
            hi = t[:, HALF:COLS]                # column slice  <-- the defect
            a = pl.col_expand_mul(lo, w)        # lo * w
            b = pl.col_expand_mul(hi, w)        # hi * w
            out = pl.assemble(out, pl.concat(a, b), [0, 0])
        return out
    return k


print(f"{'ROWS':>5}  {'max_abs_err':>12}  {'wrong elems':>14}  {'wrong rows':>12}  verdict")
print("-" * 68)
fails = []
for rows in (1, 2, 4, 8, 16):
    torch.manual_seed(0)
    x = torch.randn(rows, COLS, dtype=torch.float32)
    w = torch.randn(1, HALF, dtype=torch.float32)
    out = torch.zeros(rows, COLS, dtype=torch.float32)
    build(rows)(x, w, out, config=cfg)
    ref = x * torch.cat([w, w], dim=1)          # golden: every column scaled by w
    err = (out - ref).abs()
    bad = int((err > 1e-5).sum())
    badr = int((err.max(dim=1).values > 1e-5).sum())
    ok = bad == 0
    if not ok:
        fails.append(rows)
    print(f"{rows:>5}  {float(err.max()):>12.6f}  {bad:>6}/{rows*COLS:<7}  {badr:>5}/{rows:<6}  "
          f"{'PASS' if ok else '*** WRONG ***'}")

print()
if fails:
    print(f"REPRODUCED: column-slicing a multi-row tile is miscompiled for ROWS = {fails}.")
    print("ROWS = 1 passes (source and destination addresses coincide), which is exactly why")
    print("a single-row RoPE is unaffected while a 16-row one is destroyed.")
    # Show the signature on the largest failing case.
    rows = fails[-1]
    torch.manual_seed(0)
    x = torch.randn(rows, COLS, dtype=torch.float32)
    w = torch.randn(1, HALF, dtype=torch.float32)
    out = torch.zeros(rows, COLS, dtype=torch.float32)
    build(rows)(x, w, out, config=cfg)
    print(f"\nSignature (ROWS={rows}): out[0, 64:68] should be x[0, 64:68]*w[0,0:4]; the dense")
    print("repack instead leaves it holding row 1's LOWER half:")
    print(f"  device out[0, 64:68]      = {[round(v, 4) for v in out[0, HALF:HALF+4].tolist()]}")
    print(f"  correct  x[0,64:68]*w[:4] = "
          f"{[round(v, 4) for v in (x[0, HALF:HALF+4] * w[0, 0:4]).tolist()]}")
    print(f"  x[1, 0:4]*w[0:4]          = "
          f"{[round(v, 4) for v in (x[1, 0:4] * w[0, 0:4]).tolist()]}   <-- what actually lands")
    raise SystemExit(1)
print("NOT reproduced.")
```

</details>

## Blast radius

`t[:, a:b]` on a multi-row tile is a routine DSL construct. Any kernel doing it is silently wrong,
with no signal — and it is used identically by the *working* K path in the very same kernel, which is
what made this so hard to see.

Measured consequence in qwen3-14b's fused decode, grading against the un-fused kernel (a different
kernel, and the one that passes the serving e2e) on a fixture where every batch row shares one
`seq_len`:

| | `all_q_padded` vs a host oracle | bad rows | err@6 | argmax@6 |
|---|---|---|---|---|
| as-is | **5.765625** | **640/640** | 1.42 | 13/16 |
| with the Q RoPE forced to slice one row at a time | **0.000977** (a bf16 quantum) | **0/640** | 0.084 | **16/16** |

qwen3-14b's **serving e2e fails**. pypto-lib previously REVERTED the whole fused-decode feature
(pypto-lib#703, "Revert Qwen3 decode changes causing incorrect serving output") — this is one of the
reasons why.

## Suggested fix

The bad `MemRef` is already present at `29_after_InitMemRef`, i.e. **before** `MemoryReuse` and
`AllocateMemoryAddr` — so this is not a memory-reuse aliasing bug like #2007; the slice's shape is
mis-encoded at construction. Options:

1. **Give the sliced tile the parent's row pitch** (`pad = cols_parent - cols_slice`), making it a
   true strided view. `textract` then becomes a no-op and nothing is repacked. This looks like the
   right fix.
2. Or allocate a **genuinely disjoint** destination, so the strided -> dense repack is out-of-place.
3. Failing either, **reject it** — the precedent is `pto.alloc_multi_tile`, which refuses slot
   layouts that "would silently overlap". Silent wrong data is the worst possible outcome.

## Ruled out

Each was tested and falsified:

- **Not #2003 / #2007 / #2006** — reproduces with all three fixes applied.
- **Not `MemoryReuse`** — the mis-encoded MemRef exists before that pass runs.
- **Not UB over-subscription** — `fa_fused`'s report says Vec 98.5 KB / 184 KB = 53.5 %.
- **Not float reassociation, not the reduction tile shape, not the publish mechanism, not the
  fa_fused task structure** — each removed independently in the real kernel; none moved the error.
- **Not the K path** — proven bug-neutral by a Q/K bisect (unfolding K alone changes nothing;
  unfolding Q alone fixes it).

### PR DESCRIPTION
**Pull request by luohuan19**
Fixes [#2010](https://github.com/hw-native-sys/pypto/issues/2010)

`t[:, a:b]` on a multi-row tile feeding a `col_expand_*` op silently returned wrong data. `ROWS = 1` was correct; every `ROWS >= 2` case was corrupted, with no error, no NaN.

## Root cause

`tile.slice` inherits its source's memory, so the slice's result buffer is a **dense `[rows, slice_cols]` window sitting inside the still-live source**. The `pto.tcolexpand*` family cannot read a `pto.subview` operand, so codegen lazily materializes the slice with a `pto.textract` into that buffer — repacking *strided* (source row pitch) into *dense* (slice row pitch) **on top of its own input**.

Only row 0 survived, because its dense destination happens to equal its source address. That is exactly why a single-row slice was unaffected while a 16-row one was destroyed.

Generated `.pto` before the fix (`[16, 128]` f32 column-sliced into halves):

```mlir
%t   = pto.alloc_tile addr = 0     rows=16, cols=128     // source, still live
%lo  = pto.alloc_tile addr = 0     rows=16, cols=64      // SAME base as %t
%hi  = pto.alloc_tile addr = 256   rows=16, cols=64      // %t + 256 B
pto.textract ins(%t, 0,  0) outs(%lo)   // strided -> dense repack ON TOP OF %t
pto.textract ins(%t, 0, 64) outs(%hi)   // reads the already-trashed %t, overwrites %lo
```

## Fix

`CanonicalizeTileSlice` already rewrote *dynamic-offset* Vec slices into a `tile.extract` with a fresh buffer (#1640), but its guard tested the wrong property. A const offset only guarantees the destination **address** is right; the destination **layout** must match too, and it does so only for a **contiguous** window — a single row, or one spanning every column.

`IsDynamicSliceOffset` is replaced by `MaterializationCorruptsSource`, which rewrites whenever the lazy textract would not be an identity copy. Contiguous const-offset slices are still left untouched, so they keep sharing the source buffer rather than paying for a duplicate allocation.

Two supporting changes close the same hole from the other side:

* **`tile.extract` is now registered `not_inplace_safe()`.** Without it, `MemoryReuse` coalesced the freshly allocated extract buffer straight back onto the source it reads from, restoring the in-place repack the rewrite exists to avoid. (Verified: with the pass fix alone, the device still returned wrong data.)
* **The codegen materialization path hard-fails on a non-contiguous window** (`INTERNAL_CHECK`) instead of emitting the source-corrupting repack, so any slice that escapes the pass (e.g. a 4-arg slice carrying `valid_shape`) is rejected loudly rather than producing silently wrong data — option 3 in the issue.

## Note: the issue's repro also needs the `tile.concat` fix

The repro in #2010 trips **two independent defects**. Isolating them on device (`[16, 128]`, Ascend 910B):

| Variant | This PR | |
| --- | --- | --- |
| Column slices, **no** concat | **PASS** (was wrong for every `ROWS >= 2`) | #2010 itself — fixed here |
| Concat, **no** column slice anywhere | still wrong | needs the separate `tile.concat` `not_inplace_safe()` fix |
| The issue's repro (both) | needs both | green once both land |

The second defect is `pto.tconcat` writing its wider-pitch `[R, C0+C1]` destination onto one of its own live inputs; it is already fixed on the `fix/concat-not-inplace-safe` branch, so it is deliberately **not** duplicated here. With both changes applied together, the issue's repro passes for every `ROWS` (max|diff| = 0).

## Verification

* **Device (Ascend 910B)**: a `[16, 128]` tile column-sliced into halves through `col_expand_mul` now matches the host oracle exactly (max|diff| = 0 over all rows), where it previously diverged on 11 of 16 rows.
* **Device ST, no regressions from `not_inplace_safe`**: matmul + auto_tile_matmul (39), memory_reuse + memory_planner + mat_slice_to_left (6), broadcast + col_expand_arith (28), concat + assemble + batch_matmul + gemv (23) — all pass.
* **Unit tests**: 7043 passed.

## Tests added

* `tests/ut/ir/transforms/test_canonicalize_tile_slice.py` — the column slice of a multi-row tile is rewritten; the boundary cases that must stay untouched (single-row `[5,0]`, full-width multi-row `[16,0]`, full-shape `[0,0]`) are pinned.
* `tests/ut/codegen/test_pto_codegen_ops.py` — asserts the `pto.textract` destination is allocated at a **different address** than its source, not merely a different SSA name. The pre-fix codegen already emitted a distinct destination SSA; it was the *address* that made it corrupt, so this is the assertion that actually pins the bug.
* `tests/st/runtime/ops/test_broadcast.py` — on-device regression parametrized over `rows = [1, 2, 16]`, isolating the "row 0 always survived" symptom.

### CODE REVIEW

File: src/backend/common/pto_ops_shared.cpp [hunk 1]
```diff
@@ -405,6 +405,28 @@ std::string MaterializeSubviewOperandIfNeeded(const ir::ExprPtr& expr, codegen::
...
+  // that pass (e.g. it carries a valid_shape / drop_dims and is not a plain
+  // window). Fail loudly rather than emit silently-wrong data.
+  //
+  // Note the check reads the *immediate* source's cols; for a slice of a slice
+  // the effective pitch is the root tile's. CanonicalizeTileSlice peels such
+  // chains, so a non-contiguous chain never reaches this path.  source_cols == 0
+  // means the source columns were not statically known — stand down rather than
+  // reject a shape we cannot reason about.
+  INTERNAL_CHECK_SPAN(mat->source_cols == 0 || mat->view_rows == 1 || mat->view_cols == mat->source_cols,
+                      expr->span_)
```
[the remark this answers is not in the document]
Comment (luohuan19) [reply src/backend/common/pto_ops_shared.cpp:439]: Good catch — confirmed reachable and fixed in `14a98407`.

The pass predicate does cover dynamic offsets (`MaterializationCorruptsSource = IsDynamicSliceOffset || !IsContiguousWindow`), but `ParseCanonicalSlice` only collects plain 3-arg windows, so a slice carrying `valid_shape` / `drop_dims` escapes it — a rank-reducing `t[row]` with a dynamic `row` reached the lazy materialization with `view_rows == 1`, passed the contiguity check, and was still emitted onto the source's row 0.

The guard now also requires a const offset: `SubviewMaterializationInfo::const_offset` is set from the offset tuple in the `tile.slice` codegen and checked in `MaterializeSubviewOperandIfNeeded`, so such a slice hard-fails instead of silently corrupting its source. Strictly a tightening — slices the pass rewrites never reach this path, and const-offset slices are unaffected. New UT `test_dynamic_offset_slice_escaping_the_pass_into_col_expand_is_rejected` pins it; full UT suite green (7044 passed).

**Review by lyfne123 — APPROVED** [review]
